breakpad-symbols/src/http.rs

```diff
@@ -181,6 +182,90 @@ fn commit_cache_file(mut temp: NamedTempFile, final_path: &Path, url: &Url) -> i
     Ok(())
 }
 
+/// The result of a lookup by code_file/code_identifier against a symbol
+/// server.
+#[derive(Debug, Clone)]
+pub struct DebugInfoResult {
+    debug_file: String,
+    debug_identifier: DebugId,
+}
+
+/// Given a vector of symbol urls and a module with a code_file and code_identifier,
+/// this tries to request a symbol file using the code file and code identifier.
+///
+/// `<code file>/<code identifier>/<code file>.sym`
+///
+/// If the symbol server returns an HTTP 302 redirect, the Location header will
+/// have the correct download API url with the debug file and debug identifier.
+///
+/// This is supported by tecken
+///
+/// This returns a DebugInfoResult with the new debug file and debug identifier
+/// or None.
+async fn lookup_debug_info_by_code_info(
+    symbol_urls: &Vec<Url>,
+    module: &(dyn Module + Sync),
+) -> Option<DebugInfoResult> {
+    let lookup_path = match code_info_breakpad_sym_lookup(module) {
+        Some(value) => value,
+        _ => return None,
+    };
+
+    for base_url in symbol_urls {
+        let url = base_url.join(&lookup_path).ok()?;
```

Comment:
If `base_url.join()` here fails we'll return from the function which is probably not what you want. To avoid using `continue` and calls to `unwrap()` in the code past this point I propose doing something like this:
```rust
for base_url in symbol_urls {
  if let Some(result) = individual_lookup_debug_info_by_code_info(lookup_path) {
    return result;
  }
}
```
Then you structure `individual_lookup_debug_info_by_code_info()` to make extensive use of the comma operator to return `None` when it cannot proceed. For example:
```rust
fn individual_lookup_debug_info_by_code_info(lookup_path: &str) -> Option<DebugInfoResult> {
  let url = base_url.join(&lookup_path).ok()?;
  let no_redirects_client = Client::builder().redirect(redirect::Policy::none()).build().ok()?;
  let response = no_redirects_client.get(url.clone()).send().await.ok()?;
  ...
}
```
Basically at every step where you have a function that must succeed, turn the result into an `Option<T>` (if it's not already) and then use the question mark operator to proceed or return `None`. Note how I use the `Result::ok()` function to turn a `Result<T>` into an `Option<T>` where the error part of the latter becomes a `None`. This turns all the control flow into easy-to-read straight line code.

Reply:
That's sooooo much better! Thank you!